pkg/schema/definitions/converter.go

```diff
@@ -0,0 +1,308 @@
+		return definitions, err
+	}
+	return definitions, nil
+}
+
+func convertJSONSchemaProps(props *apiextv1.JSONSchemaProps, path proto.Path, definitions map[string]definition) (definitionField, error) {
+	if props.Type != "object" && props.Type != "array" {
+		return convertJSONSchemaPropsPrimitive(props), nil
+	}
+
+	if props.Type == "array" {
+		return convertJSONSchemaPropsArray(props, path, definitions)
+	}
+
+	if len(props.Properties) > 0 {
+		return convertJSONSchemaPropsObject(props, path, definitions)
+	}
+
+	return convertJSONSchemaPropsMap(props, path, definitions)
+}
+
+func convertJSONSchemaPropsObject(props *apiextv1.JSONSchemaProps, path proto.Path, definitions map[string]definition) (definitionField, error) {
+	field := definitionField{
+		Description: props.Description,
+		Type:        path.String(),
+	}
+
+	// CRDs don't support references yet, but we guard against recursive
+	// lookups to be safe
+	if _, ok := definitions[path.String()]; ok {
+		return field, nil
+	}
+
+	def := definition{
+		Type:           path.String(),
+		Description:    props.Description,
+		ResourceFields: map[string]definitionField{},
+	}
+
+	requiredSet := make(map[string]struct{})
+	for _, name := range props.Required {
+		requiredSet[name] = struct{}{}
+	}
+
+	for name, prop := range props.Properties {
+		subField, err := convertJSONSchemaProps(&prop, path.FieldPath(name), definitions)
+		if err != nil {
+			return definitionField{}, err
+		}
+
+		_, required := requiredSet[name]
+		subField.Required = required
+		def.ResourceFields[name] = subField
+	}
+
+	definitions[path.String()] = def
+
+	return field, nil
+}
+
+func convertJSONSchemaPropsPrimitive(props *apiextv1.JSONSchemaProps) definitionField {
+	return definitionField{
+		Description: props.Description,
+		Type:        getPrimitiveType(props.Type),
+	}
+}
+
+func convertJSONSchemaPropsArray(props *apiextv1.JSONSchemaProps, path proto.Path, definitions map[string]definition) (definitionField, error) {
+	field := definitionField{
+		Description: props.Description,
+		Type:        "array",
+	}
+	item := getItemsSchema(props)
+	if item == nil {
+		return definitionField{}, fmt.Errorf("array %q must have at least one item", path.String())
+	}
+
+	subField, err := convertJSONSchemaProps(item, path, definitions)
+	if err != nil {
+		return definitionField{}, err
+	}
+
+	field.SubType = subField.Type
+
+	return field, nil
+}
+
+func convertJSONSchemaPropsMap(props *apiextv1.JSONSchemaProps, path proto.Path, definitions map[string]definition) (definitionField, error) {
+	field := definitionField{
+		Description: props.Description,
+		Type:        "map",
+	}
+	if props.AdditionalProperties != nil && props.AdditionalProperties.Schema != nil {
+		subField, err := convertJSONSchemaProps(props.AdditionalProperties.Schema, path, definitions)
+		if err != nil {
+			return definitionField{}, err
+		}
+		field.SubType = subField.Type
+	} else {
+		// Create the object in the definitions (won't recurse)
```

Comment:
I have updated the comment because I was also confused looking at this this morning. It won't recurse because by the point we know `props` is a map, we already figured out that it doesn't have any properties, so it can't recurse on those properties.